### ISSUE TITLE
Possibility to define own preferences

### ISSUE THREAD

**Issue by mauthi** [laravel-enso/Enso#17]
I saw that there are some preferences stored in database.
It would be nice if you can add to the documentation (if this is implemented):
- how to access them in JS
- how to update them in JS
- how to add new preferences

Practical usecase from my side:
- I have a card component und I would like to store for every user if the card component is open or collapsed per default (and if the user collapses it the value in database should be updated)

Maybe there is an easy solution but I tried for 2 hours and didn't find anything...

**Comment by aocneanu** [comment]
It is implemented, but  right now we only save global preferences (language & theme).
You can check the PreferencesController@setLocalPreferences fom laravel-enso/core. Tomorrow I will post again with all the relevant details. And yes, we need to document this.

**Comment by aocneanu** [comment]
Unfortunately the code needs some refactor to proper handle preferences.

(there are some global prefs that are not used anymore; in a few days we will have a new DataTable package ready, that won't depend on external libs and this will impact preferences too; the preferences for dashboard are set under a wrong key and so on)

I will start building a new store module for this in the sometime next week and let you know when it's done.

But until then, the story:

We have a `resources/preferences.json` file that has the following structure:

```
{
    "global": {
          "pref1": "value",
          ....
     },
    "local": {
          "routeName": {
                 "pref1": "value"
                 ....
           },
          .....
    }
```

In this file we keep the default preferences for all users. We also have the `preferences` table that you already identified.
For a user that hasn't changed any prefs yet, the system will load the json file.
The first time that a user changes something, via `PreferencesController@setPreferences` a record will is made in the table with the custom preferences of that user. From that moment on, the system will load prefs from there.
When can also reset preferences to default using `PreferencesController@resetToDefault()`.

We choose `json` to have maximum flexibility for each route/page. 

Right now we don't have a getter to access preferences but you can use `this.$store.state.user.preferences.local`. Nevertheless,  we don't have a setter (I don't want to improvise, so we'll wait until the dedicated preferences module is done).

Now regarding your use case, take in consideration the following:

- Prefs are not sensitive info

- For the page(s) where you need to store preferences, you will want to update the user settings for each card change. And in the future you may have many pages / cards / users. This will potentially mean lots of DB hits.

- Usually, you want the page layout dependant on the device that you are accessing the page from. Maybe on a desktop you will "afford" 16 cards, on a laptop 10, but supposedly on phone you will want only the 3 most important. It's possible but not elegant to achieve this by storing preferences in the database.

Another approach that could address the above is storing layout preferences in the browser's LocalStorage. We already do that with the DataTables vue component and it's extremely easy to manage.

Cheers

**Comment by mauthi** [comment]
Thx for your answer, I will think about this.
Local storage sounds good too but in my use case I think it should be global.

I’m looking forward to new datatables package...

**Comment by aocneanu** [comment]
The new VDT is live :)

Docs soon.

**Comment by mauthi** [comment]
It would be nice if you can inform me here after you added documentation for own preferences.

**Comment by aocneanu** [comment]
Of course. It will be delayed a few more days, right now preferences update is broken anyway (by a recent PR merged in Laravel 5.5.22)
